### PR TITLE
Enable usage of batch_by_size for other code paths than just train

### PR DESCRIPTION
**Pull request by taylanbil**
In train.py, due to our custom data preparation requirements, we use `--input_shapes`

This doesn't exist if we train on gpu, or if we are in the sequence generation code path. Thus, we get a `Namespace doesn't have attribute` error. This PR:

* uses `batch_by_size` instead of `batch_by_size_tpu` if we're training on gpu.
* safely passes `None` if `args` doesn't have `input_shapes` (currently the case for `generate.py`, `interactive.py` etc)

### CODE REVIEW

File: fairseq/tasks/fairseq_task.py [hunk 1]
```diff
@@ -138,9 +138,17 @@ def get_batch_iterator(
...
-            indices, dataset.num_tokens, self.args.input_shapes
-        )
+        if getattr(self.args, 'use_gpu', True):
+            batch_sampler = data_utils.batch_by_size(
+                indices, dataset.num_tokens, max_tokens=max_tokens,
+                max_sentences=max_sentences,
+                required_batch_size_multiple=required_batch_size_multiple,
+            )
+        else:
+            batch_sampler = data_utils.batch_by_size_tpu(
```
Comment (dlibenzi) [inline fairseq/tasks/fairseq_task.py:148]: Where is this implemented?
Did I review it?

Comment (taylanbil) [reply fairseq/tasks/fairseq_task.py:148]: yeah, it was implemented a long time ago, let me find the pr.

Comment (dlibenzi) [reply fairseq/tasks/fairseq_task.py:148]: This needs to be "world aware" otherwise you will end up with issues at the end of the samples set.

Comment (taylanbil) [reply fairseq/tasks/fairseq_task.py:148]: so, [this is the pr](https://github.com/pytorch-tpu/fairseq/commit/7a9e90e66b095692ace9d583fd421b745308e014#diff-47dbda9fcad726ba8d5aa806c9af04ee) that made the `tpu` branch in this repo. But it was really first implemented in our `examples` repo, when our runner was still there. See [here](https://github.com/pytorch-tpu/examples/blob/xrt.r1.15/fairseq_train_tpu.py#L75)

Comment (taylanbil) [reply fairseq/tasks/fairseq_task.py:148]: when we moved our runner inside the `fairseq` repo, it got carried over.

Comment (taylanbil) [reply fairseq/tasks/fairseq_task.py:148]: ah here is the [actual pr](https://github.com/pytorch-tpu/examples/pull/11) that introduced `batch_by_size_tpu`

Comment (taylanbil) [reply fairseq/tasks/fairseq_task.py:148]: The classes that uses this sampler downstream are world aware. Unfortunately, due to the `.ceil` thing discussed before, we do get empty batches at the end for some cores. We definitely do always get the same number of iterations per core. All we then have to do is to drop the last batches, which we do [here](https://github.com/pytorch-tpu/fairseq/blob/tpu/train.py#L361)

**Review by dlibenzi — APPROVED** [review]

**Pull request merged**